ludwig/schema/hyperopt/scheduler.py

```diff
@@ -0,0 +1,509 @@
+
+# from ray.tune.schedulers import SCHEDULER_IMPORT
+# https://github.com/ray-project/ray/blob/137a1b12c3b31a3622fa5f721a05a64e9b559b05/python/ray/tune/schedulers/__init__.py#L28
+
+# from ray.tune.result import DEFAULT_RESULT_KEYS
+# Taken from https://github.com/ray-project/ray/blob/137a1b12c3b31a3622fa5f721a05a64e9b559b05/python/ray/tune/result.py
+TRAINING_ITERATION = "training_iteration"
+TIME_TOTAL_S = "time_total_s"
+TIMESTEPS_TOTAL = "timesteps_total"
+MEAN_ACCURACY = "mean_accuracy"
+MEAN_LOSS = "mean_loss"
+DEFAULT_RESULT_KEYS = (TRAINING_ITERATION, TIME_TOTAL_S, TIMESTEPS_TOTAL, MEAN_ACCURACY, MEAN_LOSS)
+
+# from ray.tune.result import DEFAULT_METRIC
+RAY_TUNE_DESULT_DEFAULT_METRIC = "_metric"
+# ----------------------------------------------------------------------------------------------------------------------
+
+scheduler_config_registry = Registry()
+
+
+def register_scheduler_config(name: str):
+    def wrap(scheduler_config: BaseSchedulerConfig):
+        scheduler_config_registry[name] = scheduler_config
+        return scheduler_config
+
+    return wrap
+
+
+@dataclass
+class BaseSchedulerConfig(schema_utils.BaseMarshmallowConfig, ABC):
+    """Base class for schedulers.
+
+    Not meant to be used directly.
+    """
+
+    type: str
+    """Name corresponding to a scheduler in `ludwig.schema.hyperopt.scheduler.scheduler_registry`.
+       Technically mutable, but attempting to load a derived scheduler with `type` set to a mismatched value will
+       result in a `ValidationError`."""
+
+
+# Field aliases to cut down on code reuse:
+def metric_alias(default=None):
+    return schema_utils.StringOptions(
+        options=list(DEFAULT_RESULT_KEYS) + [RAY_TUNE_DESULT_DEFAULT_METRIC],
+        default=default,
+        description=(
+            "The training result objective value attribute. Stopping procedures will use this attribute. If None but a "
+            "mode was passed, the ray.tune.result.DEFAULT_METRIC will be used per default."
+        ),
+    )
+
+
+def time_attr_alias(default=TRAINING_ITERATION):
+    return schema_utils.StringOptions(
+        options=list(DEFAULT_RESULT_KEYS),
+        default=default,
+        allow_none=False,
+        description=(
+            "A training result attr to use for comparing time. Note that you can pass in something non-temporal such as"
+            " training_iteration as a measure of progress, the only requirement is that the attribute should increase "
+            "monotonically."
+        ),
+    )
+
+
+def max_t_alias(default=100):
+    return schema_utils.PositiveInteger(
+        default=default,
+        description=(
+            "max time units per trial. Trials will be stopped after max_t time units (determined by time_attr) have "
+            "passed."
+        ),
+    )
+
+
+@dataclass
+class CommonSchedulerOptions:
+    time_attr: str = time_attr_alias()
+
+    metric: Optional[str] = metric_alias()
```

Comment:
@connor-mccorm addressing your comment again (https://github.com/ludwig-ai/ludwig/pull/2346#discussion_r1023370744), I am fine to change `metric` and `mode` to not be `Optional` - I just want to double confirm that removing `None` as an allowed value is not a breaking change.

Reply:
@connor-mccorm Yes, the way `metric` and `mode` are currently configured, the default is `None` regardless of whether `Optional` is there or not because `StringOptions` will set the JSON schema appropriately. But for consistency if the field supports `None` we should have the type annotation be `Optional` or `Union[str, None]`.

Reply:
@ksbrar @connor-mccorm my understanding here is that if we pass `metric` and `mode` at the top level (to RayTune), then RayTune will automatically propagate `metric` and `mode` to the scheduler and search algorithm. 
In our current implementation, we set this when we call `ray.tune.run()`, so I think it's None is a valid option for mode/metric at the Scheduler level

Reply:
In our Ludwig implementation, we use `goal` in the top level hyperopt config to map to a `mode`, and `metric` from the top level hyperopt config. It looks like we default to `loss` for metric, so I think this is definitely ok and probably the right thing to do :)

Reply:
@arnavgarg1 Okay, so keeping `None` for both is current behavior and works if I understand you correctly. Great!

That being said, I think we should remove `None` where we can as these schemas harden over time. It sounds like for now `mode` should definitely support `None` given the interaction with `goal`. We could set the default `metric` to `loss` and remove `None` as a valid option for it. WDYT?

Reply:
Sounds good to me!

Reply:
actually I take that back about `metric` - Ray explicitly supports `None` as an important code path. So I will keep things as they are with no further changes (`None` supported by both, `Optional` remains on both annotations - CC @connor-mccorm @arnavgarg1 ).